File: mobile/android/focus-android/app/src/main/java/org/mozilla/focus/fragment/FirstrunFragment.java
/* -*- Mode: Java; c-basic-offset: 4; tab-width: 20; indent-tabs-mode: nil; -*-
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
package org.mozilla.focus.fragment;

import android.content.Context;
import android.graphics.drawable.TransitionDrawable;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.annotation.Nullable;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.view.ViewPager;
import android.transition.Transition;
import android.transition.TransitionInflater;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import org.mozilla.focus.R;
import org.mozilla.focus.firstrun.FirstrunPagerAdapter;
import org.mozilla.focus.telemetry.TelemetryWrapper;
import org.mozilla.focus.utils.StatusBarUtils;

public class FirstrunFragment extends Fragment implements View.OnClickListener {
    public static final String FRAGMENT_TAG = "firstrun";

    public static final String FIRSTRUN_PREF = "firstrun_shown";

    public static FirstrunFragment create() {
        return new FirstrunFragment();
    }

    private ViewPager viewPager;

    private View background;

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);

        final Transition transition = TransitionInflater.from(context).
                inflateTransition(R.transition.firstrun_exit);

        setExitTransition(transition);

        // We will send a telemetry event whenever a new firstrun page is shown. However this page
        // listener won't fire for the initial page we are showing. So we are going to firing here.
        TelemetryWrapper.showFirstRunPageEvent(0);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, @Nullable Bundle savedInstanceState) {
        final View view = inflater.inflate(R.layout.fragment_firstrun, container, false);

        view.findViewById(R.id.skip).setOnClickListener(this);

        background = view.findViewById(R.id.background);

        final FirstrunPagerAdapter adapter = new FirstrunPagerAdapter(container.getContext(), this);

        viewPager = (ViewPager) view.findViewById(R.id.pager);
        viewPager.setFocusable(true);

        viewPager.setPageTransformer(true, new ViewPager.PageTransformer() {
            @Override
            public void transformPage(View page, float position) {
                page.setAlpha(1 - (0.5f * Math.abs(position)));
            }
        });
        viewPager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
            @Override
            public void onPageSelected(int position) {
                TelemetryWrapper.showFirstRunPageEvent(position);
            }

            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {}

            @Override
            public void onPageScrollStateChanged(int state) {}
        });

        viewPager.setClipToPadding(false);
        viewPager.setAdapter(adapter);
        viewPager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
            @Override
            public void onPageSelected(int position) {
                final TransitionDrawable drawable = (TransitionDrawable) background.getBackground();

                if (position == adapter.getCount() - 1) {
                    drawable.startTransition(200);
                } else {
                    drawable.resetTransition();
                }
            }

            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {}

            @Override
            public void onPageScrollStateChanged(int state) {}
        });

        final TabLayout tabLayout = (TabLayout) view.findViewById(R.id.tabs);
        tabLayout.setupWithViewPager(viewPager, true);

        return view;
    }

    @Override
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.next:
                viewPager.setCurrentItem(viewPager.getCurrentItem() + 1);
                break;

            case R.id.skip:
                finishFirstrun();
                TelemetryWrapper.skipFirstRunEvent();
                break;

            case R.id.finish:
                finishFirstrun();
                TelemetryWrapper.finishFirstRunEvent();
                break;

            default:
                throw new IllegalArgumentException("Unknown view");
        }
    }

    private void finishFirstrun() {
        final FragmentManager fragmentManager = getActivity().getSupportFragmentManager();

        PreferenceManager.getDefaultSharedPreferences(getContext())
                .edit()
                .putBoolean(FIRSTRUN_PREF, true)
                .apply();

        fragmentManager
                .beginTransaction()
                .remove(this)
                .commit();

        final UrlInputFragment inputFragment = (UrlInputFragment) fragmentManager.findFragmentByTag(UrlInputFragment.FRAGMENT_TAG);
        if (inputFragment != null) {
            inputFragment.showKeyboard();
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        StatusBarUtils.getStatusBarHeight(background, new StatusBarUtils.StatusBarHeightListener() {
            @Override
            public void onStatusBarHeightFetched(int statusBarHeight) {
                background.setPadding(0, statusBarHeight, 0, 0);
            }
        });
    }
}

File: mobile/android/focus-android/app/src/main/java/org/mozilla/focus/fragment/FirstrunFragment.kt
/* -*- Mode: Java; c-basic-offset: 4; tab-width: 20; indent-tabs-mode: nil; -*-
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
package org.mozilla.focus.fragment

import android.content.Context
import android.os.Bundle
import android.transition.TransitionInflater
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.preference.PreferenceManager
import androidx.viewpager.widget.ViewPager
import org.mozilla.focus.GleanMetrics.Onboarding
import org.mozilla.focus.R
import org.mozilla.focus.databinding.FragmentFirstrunBinding
import org.mozilla.focus.ext.requireComponents
import org.mozilla.focus.firstrun.FirstrunPagerAdapter
import org.mozilla.focus.state.AppAction
import org.mozilla.focus.telemetry.TelemetryWrapper
import org.mozilla.focus.utils.StatusBarUtils
import kotlin.math.abs

class FirstrunFragment : Fragment(), View.OnClickListener {

    private var _binding: FragmentFirstrunBinding? = null
    private val binding get() = _binding!!

    override fun onAttach(context: Context) {
        super.onAttach(context)

        val transition = TransitionInflater.from(context).inflateTransition(R.transition.firstrun_exit)

        exitTransition = transition

        // We will send a telemetry event whenever a new firstrun page is shown. However this page
        // listener won't fire for the initial page we are showing. So we are going to firing here.
        Onboarding.pageDisplayed.record(Onboarding.PageDisplayedExtra(0))
        TelemetryWrapper.showFirstRunPageEvent(0)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        _binding = FragmentFirstrunBinding.inflate(inflater, container, false)

        setupPager()

        binding.tabs.setupWithViewPager(binding.pager, true)

        binding.skip.setOnClickListener(this)

        return binding.root
    }

    override fun onClick(view: View) {
        val currentItem = binding.pager.currentItem
        when (view.id) {
            R.id.next -> {
                binding.pager.currentItem = binding.pager.currentItem + 1
                Onboarding.nextButtonTapped.record(Onboarding.NextButtonTappedExtra(currentItem))
            }

            R.id.skip -> {
                finishFirstrun()
                Onboarding.skipButtonTapped.record(Onboarding.SkipButtonTappedExtra(currentItem))

                TelemetryWrapper.skipFirstRunEvent()
            }

            R.id.finish -> {
                finishFirstrun()
                Onboarding.finishButtonTapped.record(Onboarding.FinishButtonTappedExtra(currentItem))

                TelemetryWrapper.finishFirstRunEvent()
            }

            else -> throw IllegalArgumentException("Unknown view")
        }
    }

    @Suppress("MagicNumber")
    private fun setupPager() {
        val firstRunPagerAdapter = FirstrunPagerAdapter(requireContext(), this)
        binding.pager.apply {
            contentDescription = firstRunPagerAdapter.getPageAccessibilityDescription(0)
            isFocusable = true

            setPageTransformer(true) { page, position ->
                page.alpha = 1 - 0.5f * abs(position)
            }

            clipToPadding = false
            adapter = firstRunPagerAdapter
            addOnPageChangeListener(object : ViewPager.OnPageChangeListener {
                override fun onPageSelected(position: Int) {
                    Onboarding.pageDisplayed.record(Onboarding.PageDisplayedExtra(0))
                    TelemetryWrapper.showFirstRunPageEvent(position)

                    contentDescription =
                        firstRunPagerAdapter.getPageAccessibilityDescription(position)
                }

                override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {}

                override fun onPageScrollStateChanged(state: Int) {}
            })
        }
    }

    private fun finishFirstrun() {
        PreferenceManager.getDefaultSharedPreferences(requireContext())
            .edit()
            .putBoolean(FIRSTRUN_PREF, true)
            .apply()

        val selectedTabId = requireComponents.store.state.selectedTabId
        requireComponents.appStore.dispatch(AppAction.FinishFirstRun(selectedTabId))
    }

    override fun onResume() {
        super.onResume()
        StatusBarUtils.getStatusBarHeight(binding.background) { statusBarHeight ->
            binding.background.setPadding(
                0,
                statusBarHeight,
                0,
                0
            )
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    companion object {
        const val FRAGMENT_TAG = "firstrun"
        const val FIRSTRUN_PREF = "firstrun_shown"

        fun create(): FirstrunFragment {
            val arguments = Bundle()

            val fragment = FirstrunFragment()
            fragment.arguments = arguments

            return fragment
        }
    }
}
